python/dgl/distributed/kvstore.py

```diff
@@ -0,0 +1,1008 @@
+        # Get shared data from server side
+        request = GetSharedDataRequest(GET_SHARED_MSG)
+        rpc.send_request(self._main_server_id, request)
+        response = rpc.recv_response()
+        for name, meta in response.meta.items():
+            shape, dtype, policy_str = meta
+            shared_data = empty_shared_mem(name+'-kvdata-', False, shape, dtype)
+            dlpack = shared_data.to_dlpack()
+            self._data_store[name] = F.zerocopy_from_dlpack(dlpack)
+            self._part_policy[name] = PartitionPolicy(policy_str, self._part_id, partition_book)
+            self._data_name_list.append(name)
+        # Get full data shape across servers
+        for name, meta in response.meta.items():
+            shape, _, _ = meta
+            data_shape = list(shape)
+            data_shape[0] = 0
+            request = GetPartShapeRequest(name)
+            # send request to all main server nodes
+            for machine_id in range(self._machine_count):
+                server_id = machine_id * self._group_count
+                rpc.send_request(server_id, request)
+            # recv response from all the main server nodes
+            for _ in range(self._machine_count):
+                res = rpc.recv_response()
+                data_shape[0] += res.shape[0]
+            self._full_data_shape[name] = tuple(data_shape)
+        # Send meta data to backup servers
+        for name, meta in response.meta.items():
+            shape, dtype, _ = meta
+            request = SendMetaToBackupRequest(name, dtype, shape)
+            # send request to all the backup server nodes
+            for i in range(self._group_count):
+                server_id = self._machine_id * self._group_count + i + 1
+                rpc.send_request(server_id, request)
+            # recv response from all the backup server nodes
+            for _ in range(self._group_count):
+                response = rpc.recv_response()
+                assert response.msg == SEND_META_TO_BACKUP_MSG
+
+    def init_data(self, name, shape, dtype, policy_str, partition_book, init_func):
+        """Send message to kvserver to initialize new data tensor and mapping this
+        data from server side to client side.
+
+        Parameters
+        ----------
+        name : str
+            data name
+        shape : list or tuple of int
+            data shape
+        dtype : dtype
+            data type
+        policy_str : str
+            partition-policy string, e.g., 'edge' or 'node'.
+        partition_book : GraphPartitionBook or RangePartitionBook
+            Store the partition information
+        init_func : func
+            UDF init function
+        """
+        assert len(name) > 0, 'name cannot be empty.'
+        assert len(shape) > 0, 'shape cannot be empty'
+        assert policy_str in ('edge', 'node'), 'policy_str must be \'edge\' or \'node\'.'
+        shape = list(shape)
+        policy_list = []
+        if self._client_id == 0:
```

Comment:
why only one client creates the data in all kvservers? in this way, the client needs to access partition book of other partitions. you can avoid this by asking all clients to create data on their local server.

Reply:
But as we disucssed before, the shape argument for init_data() is a full shape not partitioned shape right? So you must access partition book to partition shape across servers.

Reply:
we only need to get the shape of the local partition.